### PR TITLE
Feat(metadata): basic metadata api support

### PR DESCRIPTION
**Pull request by indirection42**
- No support for metadata serialization yet

### CODE REVIEW

File: xcq-extension/procedural/src/decl_extensions.rs [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+    decl_extension_inner(&item_trait)
+        .unwrap_or_else(|e| e.to_compile_error())
+        .into()
+}
+
+pub fn decl_extension_inner(item_trait: &ItemTrait) -> Result<TokenStream2> {
+    let mod_name = generate_mod_name_for_trait(&item_trait.ident);
+
+    // Assume single config associated type.
+    let mut has_config = false;
```
Comment (xlc) [inline xcq-extension/procedural/src/decl_extensions.rs:23]:
```suggestion
    let has_config = tem_trait.items.iter().any(|item| matches!(item, syn::TraitItem::Type(_)));
```

File: poc/runtime/src/xcq.rs [hunk 2]
```diff
@@ -21,38 +21,34 @@ impl xcq_extension_core::Config for ExtensionImpl {
...
-    fn balance(
-        asset: xcq_extension_fungibles::AssetIdFor<Self>,
-        who: xcq_extension_fungibles::AccountIdFor<Self>,
-    ) -> xcq_extension_fungibles::BalanceFor<Self> {
-        crate::Assets::balance(asset, who)
+impl_extensions! {
+    impl xcq_extension_core::ExtensionCore for ExtensionImpl {
+        type Config = ExtensionImpl;
+        fn has_extension(id: <Self::Config as xcq_extension_core::Config>::ExtensionId) -> bool {
+            matches!(id, xcq_extension_core::EXTENSION_ID | xcq_extension_fungibles::EXTENSION_ID)
```
[the remark this answers is not in the document]
Comment (indirection42) [reply poc/runtime/src/xcq.rs:34]: Yeah, it's just for test for now. I planned to rethink utilities provided by core extension.

File: poc/runtime/src/xcq.rs [hunk 3]
```diff
@@ -21,38 +23,34 @@ impl xcq_extension_core::Config for ExtensionImpl {
...
-    fn total_supply(asset: xcq_extension_fungibles::AssetIdFor<Self>) -> xcq_extension_fungibles::BalanceFor<Self> {
-        crate::Assets::total_supply(asset)
+
+    impl xcq_extension_fungibles::ExtensionFungibles for ExtensionImpl {
+        type Config = ExtensionImpl;
+        fn balance(
+            asset: <Self::Config as xcq_extension_fungibles::Config>::AssetId,
+            who: <Self::Config as xcq_extension_fungibles::Config>::AccountId,
+        ) -> <Self::Config as xcq_extension_fungibles::Config>::Balance {
+            crate::Assets::balance(asset, Into::<crate::interface::AccountId>::into(who))
```
[the remark this answers is not in the document]
Comment (indirection42) [reply poc/runtime/src/xcq.rs:46]: In both cases, simplifying will cause `ambiguous error`

File: xcq-extension/procedural/src/decl_extensions.rs [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+            }
+        });
+    }
+    // Add phantom data
+    variants.push(parse_quote!(
+        #[doc(hidden)]
+        __Phantom(core::marker::PhantomData<Impl>)
+    ));
+    Ok(parse_quote!(
+        #[derive(parity_scale_codec::Decode)]
```
Comment (xlc) [inline xcq-extension/procedural/src/decl_extensions.rs:82]: will this work if `parity_scale_codec` was imported with a different name? maybe we need to do something like `#xcq_extension::deps::parity_scale_codec::Decode`

File: xcq-primitives/src/lib.rs [hunk 5]
```diff
@@ -18,6 +18,7 @@ pub type XcqResult = Result<XcqResponse, XcqError>;
 pub mod metadata;
 pub mod metadata_ir;
 
-pub mod umbrella {
+pub mod deps {
```
Comment (xlc) [inline xcq-primitives/src/lib.rs:21]: add `#[doc(hidden)]` for internal use only but public items

**Review by xlc — APPROVED** [review]

**Pull request merged**